rfcs/0006-transaction.md

```diff
@@ -0,0 +1,299 @@
+
+However, it may produce massive writes on read intensive workloads when you hope to keep the read & write operations consistent, because `GetForUpdate()` simply transforms an read operation into a write operation, rewriting the key/value pair into WAL again. Also, some researches found that manual `GetForUpdate()` often introduces unexpected hard to diagnosis bugs, as described in the "2.2 Why Serializability" section of the [Serializable Snapshot Isolation in PostgreSQL](https://arxiv.org/pdf/1208.4179) paper:
+
+> Given the existence of these techniques, one might question the
+> need to provide serializable isolation in the database: shouldn’t users
+> just program their applications to handle the lower isolation level?
+> (We have often been asked this question.) Our view is that providing
+> serializability in the database is an important simplification for
+> application developers, because concurrency issues are notoriously
+> difficult to deal with. Indeed, SI anomalies have been discovered
+> in real-world applications [14]. The analysis required to identify
+> potential anomalies (or prove that none exist) is complex and is
+> likely beyond the reach of many users. In contrast, serializable
+> transactions offer simple semantics: users can treat their transactions
+> as though they were running in isolation.
+
+As above, I think it's better to finally have SSI support in the transaction feature, we could consider to implement the Snapshot Isolation first, and then implement the SSI support after it.
+
+We'll cover the implementation details about the SI and SSI in the later section.
+
+## Other Systems
+
+Before diving into the details of the transaction implementation in SlateDB, let's first review the transaction implementation of other LSM-tree based storage engines, to understand the possible approaches to implement the transaction feature in SlateDB.
+
+### RocksDB
+
+RocksDB supports both optimistic and pessimistic transactions, but we'll focus on the optimistic transaction here since it's more widely used in practice.
+
+The key idea of RocksDB's optimistic transaction is simple: track the sequence number when transaction started, and check whether the modified keys are modified by others after the transaction started during commit.
+
+The conflict checking in RocksDB is very straightforward:
+
+1. Track the sequence number when transaction is started in the transaction state.
+2. Track the modified keys in the transaction state.
+3. When `commit()` is called, check whether the modified keys are modified by others after the transaction started.
```

Comment:
Since it only tracks modified keys, does rocksdb with optimistic transaction only support snapshot isolation and not serializable snapshot isolation?
Can this aspect be added explicitly?

Reply:
yes, they only support SI without SSI as only the writes are tracked in memtable. 

updated in 35c03d8